### PR TITLE
[data-poll-manager] do not sent data poll when there is no parent (e.g., when detached)

### PR DESCRIPTION
**Pull request by abtink**

### CODE REVIEW

**Comment by jwhui** [comment]
Do we also want to address IEEE 802.15.4 Data Request messages already in the queue?

**Comment by abtink** [comment]
Good point... I think we should...

**Comment by abtink** [comment]
New pushed commit with an updated commit log:

**[data-poll-manager] No parent, no data poll!**

The goal of this change is to ensure that data polls (IEEE 802.15.4
Data Request commands) are not sent when there is no parent (or parent
candidate) or the parent is not in valid/restoring states.

In both `DataPollManager` and `MeshForwarder` classes, new checks are
added to ensure that there is a valid parent before enqueuing/sending
a data poll message. Also in `MeshForwarder`, the MAC source and
destination addresses for a data poll frame are determined when
preparing the data poll message (in `GetDirectTransmission()`) instead
of from `SendPoll()` call.

This commit also adds two helper methods in `Mac::Frame` class for the
common operation of setting source/destination addresses on a frame using
a `Mac::Address` as input (which may contain either a short or an
extended address).

File: src/core/thread/data_poll_manager.cpp [hunk 1]
```diff
@@ -97,10 +97,14 @@ otError DataPollManager::SendDataPoll(void)
     MeshForwarder &meshForwarder = GetMeshForwarder();
     otError error;
     Message *message;
+    Neighbor *parent;
 
     VerifyOrExit(mEnabled, error = OT_ERROR_INVALID_STATE);
     VerifyOrExit(!meshForwarder.GetNetif().GetMac().GetRxOnWhenIdle(), error = OT_ERROR_INVALID_STATE);
 
+    parent = meshForwarder.GetNetif().GetMle().GetParent();
+    VerifyOrExit((parent != NULL) && parent->IsStateValidOrRestoring(), error = OT_ERROR_INVALID_STATE);
```
Comment (librasungirl) [inline src/core/thread/data_poll_manager.cpp]: The return pointer of `Mle::GetParent() `would point to either `mParent` or `mParentCandidate`, no `NULL`.

Comment (abtink) [inline src/core/thread/data_poll_manager.cpp]: Yes. I noticed about `GetParent()` but we may change it in future. The check to ensure it is not NULL is definitely safer.

File: src/core/mac/mac_frame.cpp [hunk 2]
```diff
@@ -567,6 +589,28 @@ otError Frame::SetSrcAddr(const ExtAddress &aExtAddress)
...
+    case sizeof(ShortAddress):
+        error = SetSrcAddr(aAddress.mShortAddress);
+        break;
+
+    case sizeof(ExtAddress):
+        error = SetSrcAddr(aAddress.mExtAddress);
+        break;
+
+    default:
+        error = OT_ERROR_INVALID_ARGS;
```
Comment (jwhui) [inline src/core/mac/mac_frame.cpp]: Should the assert instead?

File: src/core/mac/mac_frame.cpp [hunk 3]
```diff
@@ -418,6 +418,28 @@ otError Frame::SetDstAddr(const ExtAddress &aExtAddress)
...
+    case sizeof(ShortAddress):
+        error = SetDstAddr(aAddress.mShortAddress);
+        break;
+
+    case sizeof(ExtAddress):
+        error = SetDstAddr(aAddress.mExtAddress);
+        break;
+
+    default:
+        error = OT_ERROR_INVALID_ARGS;
```
Comment (jwhui) [inline src/core/mac/mac_frame.cpp]: Should the assert instead?

File: src/core/mac/mac_frame.cpp (hunk 2 above)
Comment (abtink) [inline src/core/mac/mac_frame.cpp]: ok

**Pull request merged**